snpsnap
=======

### Publication
_SNPsnap: a Web-based tool for identification and annotation of matched SNPs_  
Tune H. Pers*, Pascal Timshel* and Joel N. Hirschhorn

An important computational step following genome-wide association studies (GWAS) is to assess whether disease or trait-associated single-nucleotide polymorphisms (SNPs) enrich for particular biological annotations. SNP-based enrichment analysis needs to account for biases such as co-localization of GWAS signals to gene-dense and high linkage disequilibrium (LD) regions, and correlations of gene size, location and function. The SNPsnap Web server enables SNP-based enrichment analysis by providing matched sets of SNPs that can be used to calibrate background expectations. Specifically, SNPsnap efficiently identifies sets of randomly drawn SNPs that are matched to a set of query SNPs based on allele frequency, number of SNPs in LD, distance to nearest gene and gene density.

Bioinformatics, first published online October 13, 2014 doi:10.1093/bioinformatics/btu655
https://academic.oup.com/bioinformatics/article/31/3/418/2365926

### Website
See website hosted at the Broad Institute for more information.
https://data.broadinstitute.org/mpg/snpsnap
