<!DOCTYPE html>
<html lang="en">
<head>
	<title>SNPsnap</title>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="description" content="SNPsnap - matching SNPs based on genetic properties. Enhance your SNP-based enrichment analysis.">
	<meta name="author" content="Broad Institute of MIT and Harvard">
	<meta name="keywords" content="Broad Institute of MIT and Harvard, history, Genome Center, Whitehead Institute/MIT Center for Genome Research, Human Genome Project, Eric Lander, Institute of Chemistry and Cell Biology, Stuart Schreiber, Initiative for Chemical Genetics, Harvard University, Harvard medical School, MIT, Massachusetts Institute of Technology, Human Genome Project" />

	<!-- FAVICON -->
	<link rel="shortcut icon" href="img/broad_logo/BroadSeal-20140621-favicon.ico">
	<link rel="icon" type="image/png" href="img/broad_logo/favicon-32x32.png" sizes="32x32">

	<!-- Custom styles for SNPsnap -->
	<link href="css/jumbotron-narrow.css" rel="stylesheet">
	<link href="css/snpsnap.css" rel="stylesheet">

	<!-- GOOGLE FONTS -->
	<link href="http://fonts.googleapis.com/css?family=Crimson+Text" rel="stylesheet" type="text/css">

	<!-- GOOGLE ANALYTICS -->
	<script src="js/snpsnap_googleanalytics.js"></script>

	<!-- jQuery -->
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>

	<!-- Bootstrap core CSS -->
	<link href="static/css/bootstrap.min.css" rel="stylesheet">
	<!-- Bootstrap core JavaScript - must come AFTER jQuery -->
	<script src="static/js/bootstrap.min.js"></script>
</head>


<body>



	<div class="container-non-responsive">

	<div class="row">
		<div class="col-xs-6"><h1 class='allerta'>SNPsnap</h1></div>
		<div class="col-xs-6"><img class="img-responsive min-limits" src="img/broad_logo/BroadLogo.png"></div>
	</div>
	<br>
		
	<div class="header">
		<!-- <ul class="nav nav-pills pull-right"> -->
		<!-- <ul class="nav nav-pills navbar-right"> --> <!-- does not work at all! -->
		<ul class="nav nav-pills">
			<li><a href="index.html">Home</a></li>
			<li class="active"><a href="about.html">About</a></li>
			<li><a href="match_snps.html">Match SNPs</a></li>
			<li><a href="database_download.html">Download</a></li>
			<li><a href="faq.html">FAQ</a></li>
			<li><a href="documentation.html">Documentation</a></li>
			<li><a href="contact.html">Contact</a></li>
			<li><a href="feedback.html">Feedback</a></li>
		</ul>
	</div>

	<div class="row jumbotron_snpsnap" >
		<div class="col-xs-offset-1 col-xs-9">
			<h3><a target="_blank" href="http://bioinformatics.oxfordjournals.org/content/31/3/418">SNPsnap: a Web-based tool for identification and annotation of matched SNPs</a></h3>
			<p><strong>Tune H. Pers*, Pascal Timshel* and Joel N. Hirschhorn</strong></p>
			<!-- *Equal contributions -->
			<div class="text-left"> <!-- *HACK*! -->
			<p>
			An important computational step following genome-wide association studies (GWAS) is to assess whether disease or trait-associated single-nucleotide polymorphisms (SNPs) enrich for particular biological annotations. 
			SNP-based enrichment analysis needs to account for biases such as co-localization of GWAS signals to gene-dense and high linkage disequilibrium (LD) regions, and correlations of gene size, location and function. 
			The SNPsnap Web server enables SNP-based enrichment analysis by providing matched sets of SNPs that can be used to calibrate background expectations. 
			Specifically, SNPsnap efficiently identifies sets of randomly drawn SNPs that are matched to a set of query SNPs based on allele frequency, number of SNPs in LD, distance to nearest gene and gene density.
			</p>
			<!-- <div style="margin-bottom:10px"><i>A fast and easy entry point to matching and annotation of 1000 Genomes SNPs.</i></div> -->
			<!-- <p><a class="btn btn-lg btn-success" href="match_snps.html" role="button">Start now</a></p> -->
		<i><a target="_blank" href="http://bioinformatics.oxfordjournals.org/content/31/3/418">Bioinformatics, first published online October 13, 2014 doi:10.1093/bioinformatics/btu655</a></i>
		</div> <!-- END div col -->
		</div> <!-- END div text-left -->
	</div>

	<!-- Applications Note -->
	<!-- *** 31/3/418 *** -->
	<!-- Bioinformatics (2015) 31 (3): 418-420 first published online October 13, 2014 doi:10.1093/bioinformatics/btu655 -->
	<br>

	<div class="row">
		<div class="col-xs-offset-1 col-xs-9">
			<div class="panel panel-default"> <!-- START feed -->
				<div class="panel-heading">
					<h3 class="panel-title">SNPsnap main usage</h3>
				</div>
				<div class="panel-body">
					<!-- A description list -->
					<dl>
						<dt>Match input SNPs</dt>
					  		<dd>SNPsnap is designed to generate matched sets of SNPs that can be used to calibrate background expectations in SNP-based enrichment analysis.</dd>
					  	<br>
					  	<dt>Annotate input SNPs using 1000 Genomes data</dt>
					  		<dd>SNPsnap allows users to quickly retrieve 1000 Genomes population specific annotation about SNP locus bounderies, nearest genes, genes in loci, minor allele frequencies and <a href="http://www.broadinstitute.org/mpg/snpsnap/documentation.html#snp_annotation_input_snps_annotated">much more</a>.</dd>
<!-- 					SNPsnap has two main features:
					<ol>
					  <li>Match input SNPs: generate matched sets of SNPs that can be used to calibrate background expectations in SNP-based enrichment analysis.</li>
					  <li>Annotate input SNPs: quickly retrieve 1000 Genomes population specific information about loci bounderies, nearest genes and genes in loci, minor allele frequency and much more.</li>
					</ol> -->
				</div>
			</div> <!-- END feed -->
			<div class="panel panel-default"> <!-- START feed -->
				<div class="panel-heading">
					<h3 class="panel-title">Tips &amp; Tricks</h3>
				</div>
				<div class="panel-body">
					<!-- A description list -->
					<dl>
						<dt>Tip #1: fast SNP input annotation</dt>
							<dd>SNPsnap's main purpose is to match input SNPs. However, many users are often only interested in annotating their input SNPs using 1000 Genomes population specific genotypes. 
							If this is also your main goal, you can set <span class="code_settings">Number of matched SNPs</span> to a low value (e.g. <code>1</code>), which will highly speed up the runtime for your job. 
							Of course you should also make sure to enable <span class="code_settings">Annotate input SNPs</span>.
							</dd>
						<br>
						<dt>Tip #2: independence of SNPs (clumping)</dt>
							<dd>SNPsnap has a nifty extra feature to check the independence of your input SNPs. 
							This feature was originally designed to facilitate proper use of the SNP-matching results.
							However, many of our users find it useful on its own - just as a quick way to clump your SNPs using <a href="https://www.cog-genomics.org/plink2/postproc#clump" target="_blank">PLINK 1.9</a>'s algorithm.
							</dd>
					</dl>
				</div>
			</div> <!-- END feed -->

		</div>
	</div>


			

 <!-- ############## FOOTER ################### -->
<div class="footer">
	<div class="col-xs-3"><p>&copy; Broad 2014</p></div>
	<div class="col-xs-8"></div>
	<div class="col-xs-1"><img class="img-responsive footer_min_limits" src="img/broad_logo/BroadSeal.png"></div>
</div>
 <!-- end FOOTER -->


</div> <!-- /container -->


</body>
</html>

