<!DOCTYPE html>
<html lang="en">
<head>
	<title>SNPsnap</title>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="description" content="SNPsnap - matching SNPs based on genetic properties. Enhance your SNP-based enrichment analysis.">
	<meta name="author" content="Broad Institute of MIT and Harvard">
	<meta name="keywords" content="Broad Institute of MIT and Harvard, history, Genome Center, Whitehead Institute/MIT Center for Genome Research, Human Genome Project, Eric Lander, Institute of Chemistry and Cell Biology, Stuart Schreiber, Initiative for Chemical Genetics, Harvard University, Harvard medical School, MIT, Massachusetts Institute of Technology, Human Genome Project" />

	<!-- FAVICON -->
	<link rel="shortcut icon" href="img/broad_logo/BroadSeal-20140621-favicon.ico">
	<link rel="icon" type="image/png" href="img/broad_logo/favicon-32x32.png" sizes="32x32">

	<!-- Custom styles for SNPsnap -->
	<link href="css/jumbotron-narrow.css" rel="stylesheet">
	<link href="css/snpsnap.css" rel="stylesheet">

	<!-- GOOGLE FONTS -->
	<link href="http://fonts.googleapis.com/css?family=Crimson+Text" rel="stylesheet" type="text/css">

	<!-- GOOGLE ANALYTICS -->
	<script src="js/snpsnap_googleanalytics.js"></script>

	<!-- jQuery -->
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>

	<!-- Bootstrap core CSS -->
	<link href="static/css/bootstrap.min.css" rel="stylesheet">
	<!-- Bootstrap core JavaScript - must come AFTER jQuery -->
	<script src="static/js/bootstrap.min.js"></script>
</head>



<body>

<div class="container-non-responsive">

<div class="row">
	<div class="col-xs-6"><h1 class='allerta'>SNPsnap</h1></div>
	<div class="col-xs-6"><img class="img-responsive min-limits" src="img/broad_logo/BroadLogo.png"></div>
</div>
<br>
	
<div class="header">
	<ul class="nav nav-pills">
		<li><a href="index.html">Home</a></li>
		<li><a href="about.html">About</a></li>
		<li><a href="match_snps.html">Match SNPs</a></li>
		<li><a href="database_download.html">Download</a></li>
		<li><a href="faq.html">FAQ</a></li>
		<li class="active"><a href="documentation.html">Documentation</a></li>
		<li><a href="contact.html">Contact</a></li>
		<li><a href="feedback.html">Feedback</a></li>
	</ul>
</div>

	<!-- ####################### TABLE OF CONTENTS #######################  -->

<!-- 	<div class="row jumbotron_snpsnap" >
		<div class="col-xs-offset-1 col-xs-9">
			<h2>Match your SNPs with SNPsnap!</h2>
			<div style="margin-bottom:10px"><i>A fast and easy entry point to match and annotate 1000 Genome SNPs.</i></div>
			<p><a class="btn btn-lg btn-success" href="match_snps.html" role="button">Start now</a></p>
		</div>
	</div> -->

	<div class="jumbotron">
		<h4 class="text-danger">Check out any new features or bug reports on the <a href="index.html#whatsnew">"What's New?"</a></h4>
		<!-- <p class="lead"></p> -->
		<!-- <p class="text-muted">We encourage that you direct any questions or feedback to the  <a href="feedback.html">feedback formula</a> or send the <a href="contact.html">SNPsnap team</a> an email. Thank you for helping improving SNPsnap</p> -->
	</div>
	

	<h2 class="documentation_header">Table of Contents</h2>
	<dl>
		<dt class="toc_title1"><a href="#doc_methodology">Methodology</a></dt>
			<ul style="list-style-type:none">
				<li class="toc_title2"><a href="#doc_aim">Aim</a></li>
				<li class="toc_title2"><a href="#doc_data">Data</a></li>
				<li class="toc_title2"><a href="#genetic_properties">Genetic Properties used to Match SNPs</a></li>
				<li class="toc_title2"><a href="#doc_algorithm">Algorithm</a></li>
			</ul>
		<dt class="toc_title1"><a href="#doc_webserver">Webserver</a></dt>
			<ul style="list-style-type:none">
				<li class="toc_title2"><a href="#doc_user_input">User Input</a></li>
				<li class="toc_title2"><a href="#snpsnap_score">SNPsnap Score</a></li>
				<li class="toc_title2"><a href="#snpsnap_matching_bias">SNPsnap Matching Bias</a></li>
				<li class="toc_title2"><a href="#doc_clumping">SNP Clumping</a></li>
				<li class="toc_title2"><a href="#doc_output_files">Output Files</a></li>
				<li class="toc_title2"><a href="#snp_annotation_input_snps_annotated">SNP Annotations - <span class="code_files">input_snps_annotated.txt</span></a></li>
				<li class="toc_title2"><a href="#snp_annotation_matched_snps_annotated">SNP Annotations - <span class="code_files">matched_snps_annotated.txt</span></a></li>
			</ul>
	</dl>


	<!-- ####################### MAIN DOCUMENTATION #######################  -->

	<h2 class="documentation_header"><a name="doc_methodology" class="documentation_header_link">Methodology</a></h2>
	<!-- <h2 class="documentation_header">Methodology</h2> -->

	<h3><a name="doc_aim" class="documentation_header_link">Aim</a></h3>
	The SNPsnap webserver enables SNP-based enrichment analysis by providing matched sets of SNPs that can be used to calibrate background expectations. Specifically, SNPsnap efficiently identifies sets of randomly drawn SNPs that are matched to a set of query SNPs based on 
	<ul>
		<li>Minor allele frequency</li>
		<li>Number of SNPs in linkage disequilibrium (LD buddies)</li>
		<li>Distance to nearest gene</li>
		<li>Gene density</li>
	</ul>

	<!-- PRODUCTION VERSION 2 -->
	<h3><a name="doc_data" class="documentation_header_link">Data</a></h3>
	SNPsnap uses  <strong>1000 Genomes Project Phase 3</strong> variants from the <strong>three different ancestral cohorts</strong>. (<span class="snpsnap_new_feature">March 2015: SNPsnap was updated from 1000G Phase 1 to Phase 3 variants</span>).
	SNPsnap uses 1000G Project's definition of the super populations for European and East Asian (see below), but defines West Africa as a subset of the 1000G African samples. 
	Super populations are defined using the <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel">1000G panel file</a>.<br>
	Specifically our database contains biallelic, uniquely mapped SNPs derived from preprocessing the phase 3 genotype data.
	SNPsnap's database only holds <strong>common variants</strong> (>1% MAF). 
	SNPsnap contains <strong>all types of variants listed by 1000 Genomes Project</strong>: single nucleotide variants (SNPs), indels and larger structural variants (as assigned with evs-numbers).<br>
	<span class="documentation_important">SNPsnap contains SNPs located on chromosomes 1-22 and the X-chromosome</span> (<span class="snpsnap_new_feature">March 2015: X chromosome included in SNPsnap</span>).<br>

	<br>
	<span class="text-muted">
	Technical note: we preprocess all 1000G variants before building SNPsnap's database using the following QC criteria:
	<ol>
		<li>Remove SNPs flagged by PLINK as merge conflicts when merging per chromosome 1000G genotypes. See more information at <a target="_blank" href="https://www.cog-genomics.org/plink2/data#merge3">PLINK 1.9 Merge failures</a>.</li>
		<li>Remove SNPs with duplicate rsIDs.</li>
		<li>Run PLINK variant filtering:
			<br>
			<samp class="text-muted">
			plink --bfile [prefix to population specific genotypes] --maf 0.01 --hwe 10e-6 --geno 0.1 --make-bed --out [SNPsnap QC PLINK files]
			</samp>
			<br>
			see details for commando at PLINK 1.9's documentation for 
			<samp><a target="_blank" href="https://www.cog-genomics.org/plink2/filter#maf">maf</a></samp>, 
			<samp><a target="_blank" href="https://www.cog-genomics.org/plink2/filter#hwe">hwe</a></samp> and 
			<samp><a target="_blank" href="https://www.cog-genomics.org/plink2/filter#missing">geno</a></samp>.
		</li>
		<li>Remove SNPs with duplicate chromosomal coordinates.</li>
	</ol>
	</span>


	<h4>SNPsnap Super Population Definitions</h4>
	SNPsnap's super population definitions are based on the following <a href="http://www.1000genomes.org/about#ProjectSamples">1000G cohorts</a>:
	<ul>
		<li><strong>European (EUR)</strong>
			<ul>
				<li>British in England and Scotland (GBR)</li>
				<li>Finnish in Finland (FIN)</li>
				<li>Iberian populations in Spain (IBS)</li>
				<li>Toscani in Italy (TSI)</li>
				<li>Utah residents with Northern and Western European ancestry (CEU)</li>
			</ul>
		</li>
		<li><strong>East Asian (EAS)</strong>
			<ul>
				<li>Chinese Dai in Xishuangbanna, China (CDX)</li>
				<li>Han Chinese in Bejing, China (CHB)</li>
				<li>Japanese in Tokyo, Japan (JPT)</li>
				<li>Kinh in Ho Chi Minh City, Vietnam (KHV)</li>
				<li>Southern Han Chinese, China (CHS)</li>
			</ul>
		</li>
		<li><strong>West Africa (WAFR)</strong>
			<ul>
				<li>Esan in Nigeria (ESN)</li>
				<li>Yoruba in Ibadan, Nigeria (YRI)</li>
				<li>Gambian in Western Division, The Gambia (GWD)</li>
				<li>Mende in Sierra Leone (MSL)</li>
			</ul>
		</li>
	</ul>


	<h4>SNPsnap Database Summary</h4>
	After several QC steps of the 1000G data, the SNPsnap's SNP database is build. The below table summarizes the number of SNPs in the database:
	<br>

	<table class='table table-hover'>
		<thead>
			<tr>
				<th class="col-xs-3"><p class="text-left" style='font-size:115%;'>Super Population</p></th>
				<th class="col-xs-8"><p class="text-left" style='font-size:115%;'>Number of SNPs in Database</p></th>
			</tr>
		</thead>
		<tbody>
			<tr>
				<th>EUR</th>
				<td>9,535,060</td>
			</tr>
			<tr>
				<th>EAS</th>
				<td>8,433,735</td>
			</tr>
			<tr>
				<th>WAFR</th>
				<td>16,191,783</td>
			</tr>
		</tbody>
	</table>

	
	<!-- PRODUCTION VERSION 1 -->
<!-- 	<h3>Data</h3>
	SNPsnap currently uses 1000 Genomes Project Phase 1 variants from the <strong>European ancestral cohorts</strong>. Specifically our database contains 9,239,614 biallelic, uniquely mapped SNPs derived from preprocessing the phase 1 genotype data for 268 samples from the CEU, GBR and TSI populations.
	SNPsnap's database only holds <strong>common variants</strong> (>1% MAF). SNPsnap contains <strong>all types of variants listed by 1000 Genomes Project</strong>: single nucleotide variants (SNPs), indels and larger structural variants (as assigned with evs-numbers).
	<span class="text-danger">SNPsnap will soon be updated with 1000 Genomes Phase 3 variants. The update will also include more populations.</span>
 -->
	<h4>SNPsnap Gene Sets</h4>
	SNPsnap uses genes from the <a href="http://www.gencodegenes.org/">GENCODE</a> consortium downloaded via <a href="http://grch37.ensembl.org">Ensembl GRCh37 Biomart</a> (Homo sapiens genes, GRCh37.p13). (<span class="snpsnap_new_feature">March 2015: Gene set updated to GENCODE genes</span>).
	SNPsnap uses any genes within the GENCODE gene set to define the <span class="code_settings">Distance to Nearest Gene</span> and <span class="code_settings">Gene Density</span>. Please note that not all these genes are coding genes.
	The SNPsnap GENCODE gene set contains 57737 genes of which 20314 are protein coding genes. SNPsnap provides the annotation feature <span class="code_column">dist_nearest_gene_snpsnap_protein_coding</span> as the distance to the nearest protein coding gene.
	Finaly, SNPsnap uses Ensembl's mapping to HGNC symbols whenever a HGNC symbol is provided.

	<h3><a name="genetic_properties" class="documentation_header_link">Genetic Properties used to Match SNPs</a></h3>
	<ul>
	  <li ><strong>Minor allele frequency</strong>: SNPs are partitioned into minor allele frequency bins of 1-2, 2-3, ..., 49-50% strata.</li>
	  <li ><strong>LD buddies</strong>: the number of “buddy” (or "proxy") SNPs in LD at various thresholds. SNPsnap currently offers LD buddy counts for thresholds using r&#178 > 0.1, 0.2, ..., 0.9</li>
	  <li ><strong>Distance to nearest gene</strong>: the distance to the nearest 5’ start site using <a href="http://www.gencodegenes.org/">GENCODE</a> gene coordinates. If the SNP is within a gene, the distance to that gene’s start site is used. </li>
	  <li ><strong>Gene density</strong>: the number of genes in loci around the SNP, using LD (r&#178 > 0.1, 0.2, ..., 0.9) and physical distance (100, 200, ..., 1000 kb) to define loci. </li>
	</ul>
	
	The genetic properties of the SNPs in the current version of SNPsnap's database is left-skewed (see below histograms).
	<span class="text-danger">*The below plots are based on the older version of SNPsnap using 1000 Genomes Phase 1 variants from European ancestry.</span>
	<div class="row">
		<div class="col-xs-offset-1 col-xs-10">
			<img class="img-responsive" src="img/documentation/Supp_Fig_1_web_edition.png">
		</div>
	</div>


	<h3><a name="doc_algorithm" class="documentation_header_link">Algorithm</a></h3>
	<div class="row">
		<div class="col-xs-offset-1 col-xs-10">
			<img class="img-responsive" src="img/documentation/Figure_1.png">
		</div>
	</div>

	The SNPsnap algorithm identifies matched SNPs as follows:
	<dl class="dl-horizontal">
		<dt><strong>Step 1</strong></dt>
		<dd>In five uniformly spaced increments, increase the allowable deviation for each of the genetic properties, ending with the prespecified maximum allowable deviation. For each increment, identify matching SNPs, defined as SNPs with genetic properties within the allowable deviations. </dd>
		<dt><strong>Step 2</strong></dt>
		<dd>If there are at least as many matching SNPs as requested, sample without replacement the requested number of SNPs from the matching SNPs and proceed to <strong>step 5</strong>. </dd>
		<dt><strong>Step 3</strong></dt>
		<dd>If the number of matching SNPs is less than number of requested SNPs, increment the allowable deviation and return to <strong>step 1</strong>; if the maximum allowable deviation has been reached, proceed to <strong>step 4</strong>. </dd>
		<dt><strong>Step 4</strong></dt>
		<dd>Sample with replacement from the matched SNPs identified in <strong>step 1</strong>.</dd>
		<dt><strong>Step 5</strong></dt>
		<dd>Proceed to next query SNP.</dd>
	</dl>
	Incrementing the allowable deviation of the genetic properties ensures that the best matching SNPs is always used.<br>
	This allows for setting a large allowable deviation and still get the best matching SNPs. 

<!-- 	<ol>
		<li>In five uniformly spaced increments, increase the allowable deviation for each of the properties, ending with the prespecified maximum allowable deviation. For each increment, identify matching SNPs, defined as SNPs with genetic properties within the allowable deviations. </li>
		<li>If there are at least as many matching SNPs as requested, sample without replacement the requested number of SNPs from the matching SNPs and proceed to <strong>step 5</strong>. </li>
		<li>If the number of matching SNPs is less than number of requested SNPs, increment the allowable deviation and return to <strong>step 1</strong>; if the maximum allowable deviation has been reached, proceed to <strong>step 4</strong>. </li>
		<li>Sample with replacement from the matched SNPs identified in <strong>step 1</strong>.</li>
		<li>Proceed to next query SNP.</li>
	</ol>
 -->

	<h2 class="documentation_header"><a name="doc_webserver" class="documentation_header_link">Webserver</a></h2>
	<!-- <h2 class="documentation_header">Webserver</h2> -->

	<h3><a name="doc_user_input" class="documentation_header_link">User input</a></h3>
	Chromosomal coordinates of SNPs e.g. 3:20145787 for a SNP on chromosome 3 at bp 20,145,787. 
	SNPsnap uses numeric codes 1-22 for autosomal chromosomes. The X chromosome is assigned the numeric code of 23 (following PLINK's encoding). <br>
	SNPsnap also accepts rs-numbers as assigned by the 1000 Genomes Project. Please note that not all variants in the 1000 Genomes Project have been assigned a rs-number and thus only can be identified by their chromosomal coordinate. <br>
	<strong>We recommend using chromosomal identifers for easier downstream processing of SNPsnap's output. Also, SNPsnap is speed optimized for chromosomal coordinates identifier inputs</strong>. <br>
	<span class="documentation_important">SNPsnap uses genome build GRCh37/hg19 for SNP coordinates.</span> <br>
	<span class="text-muted">Technical note: all input SNPs are internally mapped to chromosomal coordinate identifiers (chr:pos). The mapping can be inspected in the file <span class="code_files">input_snps_identifer_mapping.txt</span>.</span>
	
	<h4>Input format</h4>
	SNPsnap accepts the below SNP list input format as text file or copy/paste into the browser:
		<ul>
			<li>One SNP per line</li>
			<li>SNP identifers can consists of chromosomal coordinates, rs-numbers or a mixture of the two.</li>
		</ul>
	Please note that SNPs not formatted correctly, SNPs that does not exists in the database and SNPs excluded from matching will be written to the file <span class="code_files">input_snps_excluded.txt</span>. Please inspect this file to make sure all your SNPs are formatted correctly.
	<!-- REFERENCE: http://getbootstrap.com/css/#code -->

	
	<h3><a name="snpsnap_score" class="documentation_header_link">SNPsnap score</a></h3> <!-- OBS: this link is used in report_html.py. MAKE SURE THAT THE NAMES ARE IN SYNC! -->
	SNPsnap reports two scores that serve as guidelines for selecting proper matching settings.
	<ul>
	  <li ><strong>Insufficient-matches</strong>: the percentage of input SNPs for which SNPsnap is not able to identify the required number of matched SNPs</li>
	  <li ><strong>Match-size</strong>: the percentage median number of SNPs matched for the subset of SNPs with insufficient matches</li>
	</ul>
	Note that the Match-size score is only relevant to consider if the Insufficient-matches score indicates many insufficient matches.<br>
	For each score, the score value and an ordinal scoring variable is reported to the user. The below table illustrates an example of the reported SNPsnap score.
<!-- 	<div class="row">
		<div class="col-xs-offset-3 col-xs-6">
			<img class="img-responsive" src="img/rating_v2/rating_ok.png">
		</div>
	</div> -->
	<table class='table'>
	  <thead>
	    <tr>
	      <th class="col-xs-3"><p class="text-left" style='font-size:110%;'>SNPsnap score</p></th>
	      <th class="col-xs-2"><p class="text-left" style='font-size:110%;'>Value</p></th>
	      <th class="col-xs-7"><p class="text-center" style='font-size:110%;'>Rating</p></th>
	    </tr>
	  </thead>
	  <tbody>
	    <tr>
	      <th>Insufficient-matches</th>
	      <td>34.50%</td>
	      <td><img src='img/rating_v3/rating_very_poor.png' class="img-responsive" style="max-width:420px; max-height:21px" alt="SNPsnap score image"></td>
	    </tr>
	    <tr>
	      <th>Match-size</th>
	      <td>91.83%</td>
	      <td><img src='img/rating_v3/rating_very_good.png' class="img-responsive" style="max-width:420px; max-height:21px" alt="SNPsnap score image"></td>
	    </tr>
	  </tbody>
	</table>
	SNPsnap uses genome-wide significant loci from 63 traits and diseases from the GWAS Catalog (Hindorff et al., 2009) as a reference for the scoring system.
	The ordinal scoring variable (“very poor”, “poor”, “ok”, “good”, “very good”) corresponds to the scoring quintiles devrived from SNPsnap scores of the 63 phenotypes. 
	For example, a 'Insufficient-matches' score 'very good' ranks your SNPsnap query among the top 20% scores observed for the 63 GWAS catalog phenotypes.
	Note that the scoring metrics will only be valid for similar matching settings as the default SNPsnap matching settings.
	
	<h3><a name="snpsnap_matching_bias" class="documentation_header_link">SNPsnap matching bias</a></h3>

	Matching bias is defined as the ratio between the mean of the genetic properties of the input SNPs to matched SNPs. SNPsnap reports these values as part of the matching result to provide a guideline for selecting the proper matching settings.
	<br>
	The figure below shows the matching bias for each of the genetic properties as a function of the number of requested matched SNPs (using the default matching criteria). The matching bias may be explained by the left skewness of the distribution of the <a href="#genetic_properties"> genetic properties</a>.
	Large allowable deviation of the genetic properties and a large number of requested SNPs are the primary causes for the matching bias. Bias of the matched SNPs may be reduced or eliminated by lowering one or both of these parameters.
	We recommend reducing any matched bias because it may hamper downstream genetic enrichment analysis.

	<span class="text-danger">*The below plots are based on the older version of SNPsnap using 1000 Genomes Phase 1 variants from European ancestry.</span>
	<div class="row">
		<div class="col-xs-offset-1 col-xs-10">
			<img class="img-responsive" src="img/documentation/ratio_v4_multiplot2x2_dim1200x800.png">
		</div>
	</div>

	<h3><a name="doc_clumping" class="documentation_header_link">Clumping</a></h3>
	<span class="documentation_important">The input SNPs to SNPsnap should be independent if enrichment analysis is the goal of downstream investigations</span>. Failing to input independent SNPs will lead to unintended bias in the matched SNPs, which may cause an impropper background distribution.
	SNPsnap offers users to check independence of input SNPs by selecting <span class="code_settings">Report input loci independence</span>. SNPs can be clumped based on LD and physical distance thresholds.
	SNPsnap's clumping function is a wrapper around <a href="https://www.cog-genomics.org/plink2/" target="_blank">PLINK 1.9</a> (Chang et al. GigaScience, 2015) greedy algorithm for clumping SNPs. 
	See <a href="https://www.cog-genomics.org/plink2/postproc#clump" target="_blank">PLINK 1.9's documentation</a> for details.
	<br>
	SNPsnap will report one of the following messages on the result site, dependent on the outcome of the clumping:
	<ul>
		<li><span class='bg-success'>Your input SNPs are independent</span></li>
		<li><span class='bg-danger'>Your input SNPs are <i>not</i> independent</span></li>
	</ul>

	<p><span class="text-muted">
		Technical note: 
		SNPsnap writes a temporary .assoc file (two column file with the field headers "SNP" and "P") for the input SNPs. 
		The values in the "P" column is set to a fixed value for all input SNPs.
		Since all the p-values have the same value the <strong>clumping will be based on the input order of the SNPs</strong>, which makes the index SNPs rather arbitrary. We use the following command:
		<br>
		<samp class="text-muted">
		plink --bfile [prefix to population specific genotypes] --clump [tmp assoc file] --clump-r2 [user_r2] --clump-kb [user_kb]
		</samp>
	</p></span> <!-- END span TECHNICAL NOTE -->
	<!-- clumping suffers from a nontransitivity problem -->


	<h3><a name="doc_output_files" class="documentation_header_link">Output files</a></h3>
	<ul>
	  <li ><span class="code_files">matched_snps.txt</span><br>Primary output of SNPsnap; a matrix with dimension [N_input_SNPs x (Requested number of SNPs + 1)]. The first column lists the input SNPs. Each of the following columns contains a matched set of SNPs.</li>
	  <li ><span class="code_files">input_snps_excluded.txt</span><br>File listing the input SNPs excluded from analysis. There may be two reasons for an input SNP being excluded: 1) the SNP does not exist in the SNPsnap database; 2) the SNP maps to the HLA region and the <span class="code_settings">Exclude HLA SNPs</span> is enabled. First column in the file is input SNP identifier; second column is the reason for exclusion.</li>
	  <li ><span class="code_files">input_snps_annotated.txt</span> (if <span class="code_settings">Annotate input SNPs</span> selected)<br>File containing annotation of input SNPs (see <a href="#snp_annotation_input_snps_annotated">SNP Annotations</a> for details).</li>
	  <li ><span class="code_files">input_snps_clumped.txt</span> (if <span class="code_settings">Report input loci independence</span> selected)<br>File listing the clumping of the input SNPs. First column is the index SNP identifier of the clumped locus; second column is the count of clumped SNPs; third column is a comma seperated list of clumped SNPs.</li>
	  <li ><span class="code_files">input_snps_identifer_mapping.txt</span><br>File useful for mapping to/from chromosomal coordinates and rs-number identifiers. Note that some SNPs do not have rs-numbers. In such cases the chromosomal coordinate identifier is used instead.</li>
	  <li ><span class="code_files">matched_snps_annotated.txt</span> (if <span class="code_settings">Annotate matched SNPs</span> selected)<br>File containing annotation of matched SNPs (see <a href="#snp_annotation_matched_snps_annotated">SNP Annotations</a> for details).</li>
	  <li ><span class="code_files">snpsnap_summary.txt</span><br>a summery of the input parameters to SNPsnap and the SNPsnap scores.</li>
	</ul>
	<p>All files are tab delimited with the .txt file extension and can be directly imported to Excel.</p>


	<!-- style="color: black; text-decoration: none;"  -->
	<!-- <h3><a class="documentation_header_link" name="annotations">SNP Annotations - </a><span class="code_files">input_snps_annotated.txt</span></h3> -->
	<h3><a name="snp_annotation_input_snps_annotated" class="documentation_header_link">SNP Annotations - <span class="code_files">input_snps_annotated.txt</span></a></h3>
	<!-- <h5><span class="code_files">input_snps_annotated.txt</span></h5> -->
	SNPsnap currently supports the following annotations of SNPs in the file <span class="code_files">input_snps_annotated.txt</span>:<br>
	<table class='table table-hover'>
		<thead>
			<tr>
				<th class="col-xs-1"><p class="text-left" style='font-size:115%;'>#</p></th>
				<th class="col-xs-3"><p class="text-left" style='font-size:115%;'>Column Name</p></th>
				<th class="col-xs-8"><p class="text-left" style='font-size:115%;'>Column Description</p></th>
			</tr>
		</thead>
		<tbody>
			<tr>
				<th>1</th>
				<td class="code_column">snpID</td>
				<td>Chromosomal coordinates of the SNP</td>
			</tr>
			<tr>
				<th>2</th>
				<td class="code_column">rsID</td>
				<td>Rs-number of the SNP (note that in some cases the rs-number is identical to the chromosomal coordinate)</td>
			</tr>
			<tr>
				<th>3</th>
				<td class="code_column">freq_bin</td>
				<td>Frequency bin of the SNP using 1-2, 2-3, ..., 49-50% strata</td>
			</tr>
			<tr>
				<th>4</th>
				<td class="code_column">snp_maf</td>
				<td>Minor Allele Frequency (MAF) for the SNP</td>
			</tr>
			<tr>
				<th>5</th>
				<td class="code_column">gene_count</td>
				<td>The number of genes in the locus (gene density)</td>
			</tr>
			<tr>
				<th>6</th>
				<td class="code_column">dist_nearest_gene_snpsnap</td>
				<td>Distance to the start site of nearest gene the SNP is located within. If this distance is not defined (<code>"inf"</code>), the distance to the start site of nearest gene is used. That is, <span class="code_column">dist_nearest_gene_snpsnap</span> is equal to either <span class="code_column">dist_nearest_gene</span> or <span class="code_column">dist_nearest_gene_located_within</span>. Distance is in base pairs</td>
			</tr>
			<tr>
				<th>7</th>
				<td class="code_column">dist_nearest_gene_snpsnap_protein_coding</td>
				<td>Same as <span class="code_column">dist_nearest_gene_snpsnap</span> but distance to nearest protein coding gene start site.</td>
			</tr>
			<tr>
				<th>8</th>
				<td class="code_column">dist_nearest_gene</td>
				<td>Distance to the start site of nearest gene. Distance is in base pairs</td>
			</tr>
			<tr>
				<th>9</th>
				<td class="code_column">dist_nearest_gene_located_within</td>
				<td>Distance to the start site of nearest gene the SNP is located within. Distance is in base pairs. If the SNP is not located within any genes the distance is <code>"inf"</code></td>
			</tr>
			<tr>
				<th>10</th>
				<td class="code_column">loci_upstream</td>
				<td>Upstream locus boundary</td>
			</tr>
			<tr>
				<th>11</th>
				<td class="code_column">loci_downstream</td>
				<td>Downstream locus boundary</td>
			</tr>
			<tr>
				<th>12</th>
				<td class="code_column">ID_nearest_gene_snpsnap</td>
				<td><code>Ensembl Gene ID</code> for the gene used in the calculation of <span class="code_column">dist_nearest_gene_snpsnap</span></td>
			</tr>
			<tr>
				<th>13</th>
				<td class="code_column">ID_nearest_gene_snpsnap_protein_coding</td>
				<td><code>Ensembl Gene ID</code> for the gene used in the calculation of <span class="code_column">dist_nearest_gene_snpsnap_protein_coding</span></td>
			</tr>
			<tr>
				<th>14</th>
				<td class="code_column">ID_nearest_gene</td>
				<td><code>Ensembl Gene ID</code> for the gene used in the calculation of <span class="code_column">dist_nearest_gene</span></td>
			</tr>
			<tr>
				<th>15</th>
				<td class="code_column">ID_nearest_gene_located_within</td>
				<td><code>Ensembl Gene ID</code> for the gene used in the calculation of <span class="code_column">dist_nearest_gene_located_within</span>. Note that this column may be <span class="documentation_important">empty</span> if <span class="code_column">dist_nearest_gene_located_within</span> has the value <code>"inf"</code>.</td>
			</tr>
			<tr>
				<th>16</th>
				<td class="code_column">HGNC_nearest_gene_snpsnap</td>
				<td><code>HGNC Symbol</code> for the gene used in the calculation of <span class="code_column">dist_nearest_gene_snpsnap</span>. Note that this column may be <span class="documentation_important">empty</span> if Ensembl does not provide a mapping from ENSG_ID to HGNC Symbol.</td>
			</tr>
			<tr>
				<th>17</th>
				<td class="code_column">HGNC_nearest_gene_snpsnap_protein_coding</td>
				<td><code>HGNC Symbol</code> for the gene used in the calculation of <span class="code_column">dist_nearest_gene_snpsnap_protein_coding</span>. Note that this column may be <span class="documentation_important">empty</span> if Ensembl does not provide a mapping from ENSG_ID to HGNC Symbol.</td>
			</tr>
			<tr>
				<th>18</th>
				<td class="code_column">flag_snp_within_gene</td>
				<td>Value indicates of the SNP is located within a gene. Value can be <code>True</code> or <code>False</code>.</td>
			</tr>
			<tr>
				<th>19</th>
				<td class="code_column">flag_snp_within_gene_protein_coding</td>
				<td>Value indicates of the SNP is located within a protein coding gene. Value can be <code>True</code> or <code>False</code>.</td>
			</tr>
			<tr>
				<th>20</th>
				<td class="code_column">ID_genes_in_matched_locus</td>
				<td><code>Ensembl Gene ID</code> of genes overlaping with the locus boundaries. Note that this column may be <span class="documentation_important">empty</span> if <span class="code_column">gene_count</span> has the value <code>"0"</code>.</td>
			</tr>
			<tr>
				<th>21</th>
				<td class="code_column">friends_ld01</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.1</td>
			</tr>
			<tr>
				<th>22</th>
				<td class="code_column">friends_ld02</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.2</td>
			</tr>
			<tr>
				<th>23</th>
				<td class="code_column">friends_ld03</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.3</td>
			</tr>
			<tr>
				<th>24</th>
				<td class="code_column">friends_ld04</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.4</td>
			</tr>
			<tr>
				<th>25</th>
				<td class="code_column">friends_ld05</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.5</td>
			</tr>
			<tr>
				<th>26</th>
				<td class="code_column">friends_ld06</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.6</td>
			</tr>
			<tr>
				<th>27</th>
				<td class="code_column">friends_ld07</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.7</td>
			</tr>
			<tr>
				<th>28</th>
				<td class="code_column">friends_ld08</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.8</td>
			</tr>
			<tr>
				<th>29</th>
				<td class="code_column">friends_ld09</td>
				<td>Number of LD buddies using cutoff r&#178 > 0.9</td>
			</tr>
		</tbody>
	</table>


	<!-- ################ SNP Annotations - matched_snps_annotated.txt ############### -->
	<h3><a name="snp_annotation_matched_snps_annotated" class="documentation_header_link">SNP Annotations - <span class="code_files">matched_snps_annotated.txt</span></a></h3>
	The <span class="code_files">matched_snps_annotated.txt</span> file contains the same columns as described for <span class="code_files">input_snps_annotated.txt</span> but annotating the matched SNP.<br>
	The file contains the following two additional leading columns:<br>
	<table class='table table-hover'>
		<thead>
			<tr>
				<th class="col-xs-1"><p class="text-left" style='font-size:115%;'>#</p></th>
				<th class="col-xs-3"><p class="text-left" style='font-size:115%;'>column name</p></th>
				<th class="col-xs-8"><p class="text-left" style='font-size:115%;'>column description</p></th>
			</tr>
		</thead>
		<tbody>
			<tr>
				<th>1</th>
				<td class="code_column">set</td>
				<td>Index of the matched SNP group. The number of set index is equal to the number of required matched SNPs</td>
			</tr>
			<tr>
				<th>2</th>
				<td class="code_column">input_snp</td>
				<td>The chromosomal coordinates for the input SNPs that were matched</td>
			</tr>
			<tr>
				<th>3...31</th>
				<td>...</td>
				<td>... same as for  <span class="code_files">input_snps_annotated.txt</span> but annotation is for the matched SNP</td>
			</tr>
		</tbody>
	</table>




 <!-- ############## FOOTER ################### -->
<div class="footer">
	<div class="col-xs-3"><p>&copy; Broad 2014</p></div>
	<div class="col-xs-8"></div>
	<div class="col-xs-1"><img class="img-responsive footer_min_limits" src="img/broad_logo/BroadSeal.png"></div>
</div>
 <!-- end FOOTER -->



</div> <!-- /container -->

</body>
</html>





	<!--<div class="code_output">
	</div>
	-->



<!-- ################ LIST LIKE STRUCTURE -->
<!-- 	<ol>
		<li>snpID: chromosomal coordinates of the SNP</li>
		<li>rsID: rs-number of the SNP (note that in some cases the rs-number is identical to the chromosomal coordinate) </li>
		<li>freq_bin: the frequency bin of SNP (1-2, 2-3, ..., 49-50% strata)</li>
		<li>gene_count: an alias for gene density; the number of genes in the locus</li>
		<li>dist_nearest_gene_snpsnap: distance to the start site of nearest gene the SNP is located within. If this distance is not defined (<code>"inf"</code>), the distance to the start site of nearest gene is used. That is, <code>dist_nearest_gene_snpsnap</code> is equal to either <code>dist_nearest_gene</code> or <code>dist_nearest_gene_located_within</code>. Distance is in base pairs</li>
		<li>dist_nearest_gene: distance to the start site of nearest gene. Distance is in base pairs</li>
		<li>dist_nearest_gene_located_within: distance to the start site of nearest gene the SNP is located within. Distance is in base pairs. If the SNP is not located within any genes the distance is <code>"inf"</code></li>
		<li>loci_upstream: upstream locus boundary (that is the 5' end of the locus boundary)</li>
		<li>loci_downstream: downstream locus boundary (that is the 3' end of the locus boundary)</li>
		<li>ID_nearest_gene_snpsnap: <code>Ensembl Gene ID</code> for the gene used in the calculation of <code>dist_nearest_gene_snpsnap</code></li>
		<li>ID_nearest_gene	<code>Ensembl Gene ID</code> for the gene used in the calculation of <code>dist_nearest_gene</code></li>
		<li>ID_nearest_gene_located_within	<code>Ensembl Gene ID</code> for the gene used in the calculation of <code>dist_nearest_gene_located_within</code></li>
		<li>ID_genes_in_matched_locus: <code>Ensembl Gene ID</code> of genes overlaping with the locus boundaries</li>
		<li>friends_ld01: number of LD buddies using cutoff r&#178 > 0.1</li>
		<li>friends_ld02: number of LD buddies using cutoff r&#178 > 0.2</li>
		<li>friends_ld03: number of LD buddies using cutoff r&#178 > 0.3</li>
		<li>friends_ld04: number of LD buddies using cutoff r&#178 > 0.4</li>
		<li>friends_ld05: number of LD buddies using cutoff r&#178 > 0.5</li>
		<li>friends_ld06: number of LD buddies using cutoff r&#178 > 0.6</li>
		<li>friends_ld07: number of LD buddies using cutoff r&#178 > 0.7</li>
		<li>friends_ld08: number of LD buddies using cutoff r&#178 > 0.8</li>
		<li>friends_ld09: number of LD buddies using cutoff r&#178 > 0.9</li>
	</ol> -->


<!-- 	<ol>
		<li>set: index of the matched SNP group. The number of set index is equal to the number of required matched SNPs</li>
		<li>input_snp: the input that were matched </li>
		<li>... same as for  <span class="code_files">input_snps_annotated.txt</span> </li>
	</ol> -->
