<!DOCTYPE html>
<html lang="en">
<head>
	<title>SNPsnap</title>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="description" content="SNPsnap - matching SNPs based on genetic properties. Enhance your SNP-based enrichment analysis.">
	<meta name="author" content="Broad Institute of MIT and Harvard">
	<meta name="keywords" content="Broad Institute of MIT and Harvard, history, Genome Center, Whitehead Institute/MIT Center for Genome Research, Human Genome Project, Eric Lander, Institute of Chemistry and Cell Biology, Stuart Schreiber, Initiative for Chemical Genetics, Harvard University, Harvard medical School, MIT, Massachusetts Institute of Technology, Human Genome Project" />

	<!-- FAVICON -->
	<link rel="shortcut icon" href="img/broad_logo/BroadSeal-20140621-favicon.ico">
	<link rel="icon" type="image/png" href="img/broad_logo/favicon-32x32.png" sizes="32x32">

	<!-- Custom styles for SNPsnap -->
	<link href="css/jumbotron-narrow.css" rel="stylesheet">
	<link href="css/snpsnap.css" rel="stylesheet">

	<!-- GOOGLE FONTS -->
	<link href="http://fonts.googleapis.com/css?family=Crimson+Text" rel="stylesheet" type="text/css">

	<!-- GOOGLE ANALYTICS -->
	<script src="js/snpsnap_googleanalytics.js"></script>

	<!-- jQuery -->
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>

	<!-- Bootstrap core CSS -->
	<link href="static/css/bootstrap.min.css" rel="stylesheet">
	<!-- Bootstrap core JavaScript - must come AFTER jQuery -->
	<script src="static/js/bootstrap.min.js"></script>
</head>



<body>

<div class="container-non-responsive">

<div class="row">
	<div class="col-xs-6"><h1 class='allerta'>SNPsnap</h1></div>
	<div class="col-xs-6"><img class="img-responsive min-limits" src="img/broad_logo/BroadLogo.png"></div>
</div>
<br>
	
<div class="header">
	<ul class="nav nav-pills">
		<li><a href="index.html">Home</a></li>
		<li><a href="match_snps.html">Match SNPs</a></li>
		<li><a href="faq.html">FAQ</a></li>
		<li class="active"><a href="documentation.html">Documentation</a></li>
		<li><a href="contact.html">Contact</a></li>
	</ul>
</div>

<p>
	<h2>Methodology</h2>

	<h3>Aim</h3>
	The SNPsnap webserver enables SNP-based enrichment analysis by providing matched sets of SNPs that can be used to calibrate background expectations. Specifically, SNPsnap efficiently identifies sets of randomly drawn SNPs that are matched to a set of query SNPs based on 
	<ul>
	<li>minor allele frequency</li>
	<li>number of SNPs in linkage disequilibrium (LD buddies)</li>
	<li>distance to nearest gene</li>
	<li>gene density.</li>
	   
	</ul>
	
	<h3>Genetic properties used to sample random but matched SNPs</h3>
	<ul>
	  <li ><strong>Minor allele frequency</strong>: SNPs are partitioned into minor allele frequency bins of 1-2, 2-3, ..., 49-50% strata.</li>
	  <li ><strong>LD buddies</strong>: the number of “buddy” (or "proxy") SNPs in LD at various thresholds. SNPsnap currently offer LD buddy counts for thresholds using r&#178 > 0.1, 0.2, ..., 0.9</li>9.<li>
	  <li ><strong>Distance to nearest gene</strong>: the distance to the nearest 5’ start site using Ensembl gene coordinates (Flicek et al., 2014). If the SNP is within a gene, the distance to that gene’s start site is used. </li>
	  <li ><strong>Gene density</strong>: the number of genes in loci around the SNP, using LD (r&#178 > 0.1, 0.2, ..., 0.9) and physical distance (100, 200, ..., 1000 kb) to define loci. </li>
	</ul>
	
	<h3>Algorithm</h3>
	<div class="row">
		<div class="col-xs-offset-1 col-xs-10">
			<img class="img-responsive" src="img/documentation/Figure_1.png">
		</div>
	</div>
</p>

<p>
	<h2>Webserver</h2>

	<h3>User input</h3>
	Chromosomal positions of SNPs (not rsIDs), e.g. 3:20145787 for a SNP on chromosome 3 at bp 20,145,787. <div style="color: red;">Use genome build GRCh37/hg19 coordinates.</div>
	
	
	<h3>SNPsnap score</h3>
	SNPsnap reports two scores that serve as guidelines for selecting proper matching settings.
	<ul>
	  <li ><strong>Insufficient-matches</strong>: the number of SNPs for which SNPsnap is not able to identify the required number of matched SNPs</li>
	  <li ><strong>Match-size</strong>: median number of SNPs matched for the subset of SNPs with insufficient matches</li>
	</ul>
	For each score, an ordinal scoring variable (“very poor”, “poor”, “ok”, “good”, “very good”; see example below) is reported to the user.
	<div class="row">
		<div class="col-xs-offset-1 col-xs-10">
			<img class="img-responsive" src="img/rating/rating_ok.png">
		</div>
	</div>
	SNPsnap uses genome-wide significant loci from 63 traits and diseases from the GWAS Catalog (Hindorff et al., 2009) as a reference to the scoring system.
	
	Note that the scoring metrics will only be valid for similar matching settings as the default SNPsnap matching settings.
	
	<h3> Output files</h3>
	<ul>
	  <li ><span class="code_files">matched_snps.tab</span>: matrix with input SNPs as rows and matched SNPs as columns</li>
	  <li ><span class="code_files">input_snps_excluded.tab</span>: first column is input SNP; second column is reason for exclusion, e.g. SNP does not exist in the SNPsnap SNP data base or SNP maps to HLA region (if the setting is enabled)</li>
	  <li ><span class="code_files">input_snps_annotated.tab</span>: annotation of input SNPs (see <a href="#annotations">SNP Annotations</a> for details)</li>
	  <li ><span class="code_files">matched_snps_annotated.tab</span>: annotation of matched SNPs (see <a href="#annotations">SNP Annotations</a> for details) </li>
	  <li ><span class="code_files">snpsnap_summary.txt</span>: a summery of the input parameters to SNPsnap and the SNPsnap scores</li>
	</ul>
	
	<h3><a name="annotations" style="text-decoration: none; color: black;">SNP Annotations</a></h3>
	SNPsnap currently supports the following annotations of SNPs:<br>
	<ul>
	
	<li>set</li>
	<li>input_snp</li>
	<li>snpID	</li>
	<li>rsID	</li>
	<li>freq_bin	</li>
	<li>gene_count	</li>
	<li>dist_nearest_gene_snpsnap	</li>
	<li>dist_nearest_gene	</li>
	<li>dist_nearest_gene_located_within	</li>
	<li>loci_upstream	</li>
	<li>loci_downstream	<br></li>
	<li>ID_nearest_gene_snpsnap	</li>
	<li>ID_nearest_gene	</li>
	<li>ID_nearest_gene_located_within	</li>
	<li>ID_genes_in_matched_locus	</li>
	<li>friends_ld01	</li>
	<li>friends_ld02	</li>
	<li>friends_ld03	</li>
	<li>friends_ld04	</li>
	<li>friends_ld05	</li>
	<li>friends_ld06	</li>
	<li>friends_ld07	</li>
	<li>friends_ld08	</li>
	<li>friends_ld09 </li>
	</ul>
	<!--<div class="code_output">
	</div>
	-->

</p>
	
<!--
<h2>7. FAQ</h2>
<ul>
  <li ><i>What gene annotations does SNPsnap use for column locus genes like<span class="code_output">ID_genes_in_matched_locus</span>?</i><br>
  SNPsnap uses the complete set of genes from Ensembl which covers about 48,000 genes </li>
<i>Why is there a replicated input SNPs in <span class="code_files">matched_snps.tab</span> even though I enabled <span class="code_settings">Exclude Input SNPs</span>?</i><br>  
If no matching SNPs are found, SNPsnap will replicate the input SNP to form a full row in <span class="code_files">matched_snps.tab</span>.
</ul>
-->


 <!-- ############## FOOTER ################### -->
<div class="footer">
	<div class="col-xs-3"><p>&copy; Broad 2014</p></div>
	<div class="col-xs-8"></div>
	<div class="col-xs-1"><img class="img-responsive footer_min_limits" src="img/broad_logo/BroadSeal.png"></div>
</div>
 <!-- end FOOTER -->



</div> <!-- /container -->

</body>
</html>
