<!DOCTYPE html>
<html lang="en">
<head>
	<title>SNPsnap</title>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="description" content="SNPsnap - matching SNPs based on genetic properties. Enhance your SNP-based enrichment analysis.">
	<meta name="author" content="Broad Institute of MIT and Harvard">
	<meta name="keywords" content="Broad Institute of MIT and Harvard, history, Genome Center, Whitehead Institute/MIT Center for Genome Research, Human Genome Project, Eric Lander, Institute of Chemistry and Cell Biology, Stuart Schreiber, Initiative for Chemical Genetics, Harvard University, Harvard medical School, MIT, Massachusetts Institute of Technology, Human Genome Project" />

	<!-- FAVICON -->
	<link rel="shortcut icon" href="img/broad_logo/BroadSeal-20140621-favicon.ico">
	<link rel="icon" type="image/png" href="img/broad_logo/favicon-32x32.png" sizes="32x32">

	<!-- Custom styles for SNPsnap -->
	<link href="css/jumbotron-narrow.css" rel="stylesheet">
	<link href="css/snpsnap.css" rel="stylesheet">

	<!-- GOOGLE FONTS -->
	<link href="http://fonts.googleapis.com/css?family=Crimson+Text" rel="stylesheet" type="text/css">

	<!-- GOOGLE ANALYTICS -->
	<script src="js/snpsnap_googleanalytics.js"></script>

	<!-- jQuery -->
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>

	<!-- Bootstrap core CSS -->
	<link href="static/css/bootstrap.min.css" rel="stylesheet">
	<!-- Bootstrap core JavaScript - must come AFTER jQuery -->
	<script src="static/js/bootstrap.min.js"></script>

</head>



<body>

<div class="container-non-responsive">

<div class="row">
	<div class="col-xs-6"><h1 class='allerta'>SNPsnap</h1></div>
	<div class="col-xs-6"><img class="img-responsive min-limits" src="img/broad_logo/BroadLogo.png"></div>
</div>
<br>
	
<div class="header">
	<ul class="nav nav-pills">
		<li><a href="index.html">Home</a></li>
		<li><a href="about.html">About</a></li>
		<li><a href="match_snps.html">Match SNPs</a></li>
		<li><a href="database_download.html">Download</a></li>
		<li class="active"><a href="faq.html">FAQ</a></li>
		<li><a href="documentation.html">Documentation</a></li>
		<li><a href="contact.html">Contact</a></li>
		<li><a href="feedback.html">Feedback</a></li>
	</ul>
</div>



<div class="jumbotron">
	<h1>Any Questions?</h1>
	<!-- <p class="lead"></p> -->
	<p class="text-muted">We encourage that you direct any questions or feedback to the  <a href="feedback.html">feedback formula</a> or send the <a href="contact.html">SNPsnap team</a> an email. Thank you for helping improving SNPsnap</p>
</div>


<div id="faq_overview">
	<!-- HERE GOES THE OVERVIEW OF THE FAQ -->
	<!-- This should be a LIST of link with the correct headers -->
</div>


<h4 class="faq_section">Browser Compatibility</h4>
<dl class="dl">
	<dt class="faq_question">Does SNPsnap support all browsers?</dt>
	<dd class="faq_answer">SNPsnap is fully compatible with Safari 7, Firefox 30 and Chrome 35. For the best user experience we encourage using Chrome 35 or newer. <br>
	Internet Explorer 11 is supported but may not provide the best user experience. Users using older versions of Internet Explorer (10 or older) may experience problems.</dd>
</dl>

<h4 class="faq_section">Variant Identifiers and Annotation</h4>
<dl class="dl">
	<dt class="faq_question">Will I be able to use rs-numbers as input SNP identifiers?</dt>
	<dd class="faq_answer">Yes. SNPsnap accepts rs-numbers and chromosomal coordinates (from GRCh37/hg19) as SNP identifiers. <strong>Please note</strong> that not all variants in the 1000 Genomes Project have been assigned a rs-number. When no rs-number is given in the 1000 Genomes data, SNPsnap uses the chromosomal coordinate as a substitute for the rs-number.</dd>
	<dt class="faq_question">What types of variants are included in the SNPsnap database?</dt>
	<dd class="faq_answer">All types of variants listed by 1000 Genomes Project</strong> are included: single nucleotide variants (SNPs), indels and larger structural variants. Note that SNPsnap's chromosomal coordinates identifers do not describe the type of variant, as it is the case for IMPUTE-format (e.g. chr1:100123071:D or chr1:100134769:I for insertions and deletions, respectively).</dd>
</dl>


<h4 class="faq_section">Output</h4>
<dl class="dl">
	<!-- KEEP THE BELOW FOR LATER! -->
<!-- 	<dt class="faq_question">Is the number of input loci equal to the number of matched loci?</dt>
	<dd class="faq_answer">No. SNPsnap does not match SNPs to produce the same number of matched loci as input loci. In fact, SNPsnap's macthed SNPs will likely consist of many more loci than the input number of loci because of random sampling of matched SNPs.</dd> -->
	<dt class="faq_question">What gene annotations does SNPsnap use for column locus genes like<span class="code_output">ID_genes_in_matched_locus</span>?</dt>
	<!-- <dd>SNPsnap uses the complete set of genes from the Ensembl database and covers about 48,000 genes.</dd> -->
	<dd class="faq_answer">SNPsnap uses ~60,000 genes from the GENCODE gene set and including protein coding genes, non-coding genes and pseudo genes.</dd>
	<dt class="faq_question">Why is there a replicated input SNPs in <span class="code_files">matched_snps.tab</span> even though I enabled <span class="code_settings">Exclude input SNPs in matched SNPs</span>?</dt>
	<dd class="faq_answer">In the unlikely event that no matching SNPs are found, SNPsnap will fill in the input SNP instead of the matched SNPs.</dd>
	<!-- <span class="code_files">matched_snps.tab</span> -->
</dl>


<h4 class="faq_section">Results</h4>
<dl class="dl">
	<dt class="faq_question">What are the run times for SNPsnap?</dt>
	<dd class="faq_answer">SNPsnap matching takes ~5 seconds per input SNP when requesting 10,000 SNPs (matching 100 input SNPs will take ~8 minutes to match). Run time for matching may be reduced by a factor two if less matched SNPs is requested. If any annotation option (<span class="code_settings">Annotate matched SNPs</span> or <span class="code_settings">Annotate input SNPs</span>) is enabled SNPsnap uses an additional ~2 minutes in total. If the option <span class="code_settings">Report independent loci</span> is enabled SNPsnap uses ~5 minutes for clumping 10 SNPs using default clumping thresholds (r=0.5 and kb=250). We note that the runtime for clumping is very sensitive to the clumping thresholds.</dd>
	<dt class="faq_question">For how long are my files saved on the web server?</dt>
	<dd class="faq_answer">The SNPsnap web server stores the result files for 24 hours. After this your result files will be permanently deleted.</dd>
</dl>


<!-- ####### TEMPLATE ########### -->
<!-- <dl class="dl">
	<dt class="faq_question"></dt>
	<dd class="faq_answer"></dd>
</dl> -->


 <!-- ############## FOOTER ################### -->
<div class="footer">
	<div class="col-xs-3"><p>&copy; Broad 2014</p></div>
	<div class="col-xs-8"></div>
	<div class="col-xs-1"><img class="img-responsive footer_min_limits" src="img/broad_logo/BroadSeal.png"></div>
</div>
 <!-- end FOOTER -->



</div> <!-- /container -->

</body>
</html>
