<!DOCTYPE html>
<html lang="en">
<head>
	<title>SNPsnap</title>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="description" content="SNPsnap - matching SNPs based on genetic properties. Enhance your SNP-based enrichment analysis.">
	<meta name="author" content="Broad Institute of MIT and Harvard">
	<meta name="keywords" content="Broad Institute of MIT and Harvard, history, Genome Center, Whitehead Institute/MIT Center for Genome Research, Human Genome Project, Eric Lander, Institute of Chemistry and Cell Biology, Stuart Schreiber, Initiative for Chemical Genetics, Harvard University, Harvard medical School, MIT, Massachusetts Institute of Technology, Human Genome Project" />

	<!-- FAVICON -->
	<link rel="shortcut icon" href="img/broad_logo/BroadSeal-20140621-favicon.ico">
	<link rel="icon" type="image/png" href="img/broad_logo/favicon-32x32.png" sizes="32x32">

	<!-- Custom styles for SNPsnap -->
	<link href="css/jumbotron-narrow.css" rel="stylesheet">
	<link href="css/snpsnap.css" rel="stylesheet">

	<!-- GOOGLE FONTS -->
	<link href="http://fonts.googleapis.com/css?family=Crimson+Text" rel="stylesheet" type="text/css">

	<!-- GOOGLE ANALYTICS -->
	<script src="js/snpsnap_googleanalytics.js"></script>

	<!-- jQuery -->
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>

	<!-- Bootstrap core CSS -->
	<link href="static/css/bootstrap.min.css" rel="stylesheet">
	<!-- Bootstrap core JavaScript - must come AFTER jQuery -->
	<script src="static/js/bootstrap.min.js"></script>
</head>


<body>



	<div class="container-non-responsive">

	<div class="row">
		<div class="col-xs-6"><h1 class='allerta'>SNPsnap</h1></div>
		<div class="col-xs-6"><img class="img-responsive min-limits" src="img/broad_logo/BroadLogo.png"></div>
	</div>
	<br>
		
	<div class="header">
		<!-- <ul class="nav nav-pills pull-right"> -->
		<!-- <ul class="nav nav-pills navbar-right"> --> <!-- does not work at all! -->
		<ul class="nav nav-pills">
			<li class="active"><a href="index.html">Home</a></li>
			<li><a href="about.html">About</a></li>
			<li><a href="match_snps.html">Match SNPs</a></li>
			<li><a href="database_download.html">Download</a></li>
			<li><a href="faq.html">FAQ</a></li>
			<li><a href="documentation.html">Documentation</a></li>
			<li><a href="contact.html">Contact</a></li>
			<li><a href="feedback.html">Feedback</a></li>
		</ul>
	</div>

<!-- 	<div class="row margin_bottom0">
		<div class="jumbotron col-xs-offset-1 col-xs-9">
			<h2>Match your SNPs with SNPsnap!</h2>
			<p><a class="btn btn-lg btn-success" href="match_snps.html" role="button">Start now</a></p>
		</div>
	</div> -->

	<!-- <p class="lead"></p> -->

	<!-- align="center" -->
	<div class="row jumbotron_snpsnap" >
		<div class="col-xs-offset-1 col-xs-9">
			<h2>Match your SNPs with SNPsnap!</h2>
			<div style="margin-bottom:10px"><i>A fast and easy entry point to match and annotate 1000 Genome SNPs.</i></div>
			<p><a class="btn btn-lg btn-success" href="match_snps.html" role="button">Start now</a></p>
		</div>
	</div>


<!-- TEMPORARY OUTCOMMENTED -->
<!-- 	<div class="row">
		<div class="col-xs-offset-1 col-xs-9">
			<h2>Introduction</h2>
			<p>SNPsnap is a novel tool for machting your SNPs</p>

			<h2>Why use SNPsnap?</h2>
			<p>SNPsnap is developed for Genome-wide association studies that require a background distribution to test significance of biological hypothesis</p>
		</div>
	</div> -->

	<div class="row">
		<div class="col-xs-offset-1 col-xs-9">
			<!-- <h2>What's New?</h2> -->
			<h2><a name="whatsnew" class="documentation_header_link">What's New?</a></h2>
			<!-- <div class="scrollable_news"> -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">March 10th, 2019</div>
						<h3 class="panel-title">Server load and down-time</h3>
					</div>
					<div class="panel-body">
						We frequently experience high load on the SNPsnap web-server. Although we are happy about the demand and use of SNPsnap, it unfortunately means that the server crashes from time to time.
						When the SNPsnap server has crashed, it returns the page '<code>Internal server error (code 500)</code>' after you've submitted your job. 
						If you get this error message, you will usually have to wait one or two hours (but sometimes up to 24 hours) before the server has rebooted and is online again.
						If the server crashes all running jobs are lost (and you will not receive an email notifying you about the job completion). 
						<br>
						You should only <a href="contact.html">contact us</a> if the problem persists for more than two days.
						We are currently looking into solving this problem by 1) emailing users if their jobs failed because of server overload; and 2) upgrading the SNPsnap server to better support the usage load.
						We are sorry about the inconvenience that this downtime may cause our users.
						<br>
						<strong>To help minimize the downtime, we ask our users to be mindful about the number of jobs they submit simultaneously. Do not submit multiple jobs containing tens of thousands input SNPs as this is likely to crash the server.</strong>
					</div>
				</div> <!-- END feed -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">July 30th, 2015</div>
						<h3 class="panel-title">No matching of X chromosome SNPs</h3>
					</div>
					<div class="panel-body">
						SNPsnap is not able to match X chromosome SNPs. Input X chromosome SNPs are still accepted but they will be excluded during the matching algorithm (exclusion mark: <code>SNP_genetic_properties_corrupted_in_data_base</code>).
						We are sorry about the inconvenience this may cause our users.
					</div>
				</div> <!-- END feed -->
				<!-- <div class="panel panel-default">
					<div class="panel-heading">
						<div class="text-right">July 30, 2015</div>
						<h3 class="panel-title"><span class="text-danger">BUG report</span>: missing gene annotation for X chromosome SNPs</h3>
					</div>
					<div class="panel-body">
						We have discovered a bug in SNPsnap's SNP database. All X chromosome variants are missing the gene annotation.
						This effectively means that <strong>SNPsnap is not able to match X chromosome SNPs at the current moment</strong>.
						We are working on fixing this problem and updating SNPsnap's database, but the timeline for the fix is unknown.
						Input X chromosome SNPs are still accepted but they will be excluded during the matching algorithm (exclusion mark: <code>SNP_genetic_properties_corrupted_in_data_base</code>).
						<br>
						We are sorry about the inconvenience this may cause our users.
					</div>
				</div>  -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">March 18th, 2015</div>
						<h3 class="panel-title">Download SNPsnap's Annotation Files</h3>
					</div>
					<div class="panel-body">
						SNPsnap now offers to download SNPsnap's annotation database.
						Check it out <a target="_blank" href="database_download.html">here</a>
					</div>
				</div> <!-- END feed -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right"><div class="text-danger"><strong>March 10th, 2015</strong></div></div>
						<h3 class="panel-title"><div class="text-danger">Major SNPsnap update: 1000G Phase 3 Data</div></h3>
					</div>
					<div class="panel-body">
						SNPsnap have undergone a major update. Here is the list of new features:
						<ul>
						  <li>SNPsnap is now based on <strong><a href="http://www.1000genomes.org/about#ProjectSamples">1000 Genome Phase 3</a></strong> data. The old Phase 1 data is no longer supported.</li>
						  <li>Super populations (European, EUR; East Asian, EAS; West African, WAFR) are added to allow more specific SNP matching based on your study population.</li>
						  <li>X-chromosome SNPs included.</li>
						  <li>Geneset updated to <a href="http://www.gencodegenes.org/">GENCODE</a> genes.</li>
						  <li>New SNPsnap annotation features: MAF, HGNC symbols, distance to nearest protein coding gene, flag for SNP within gene and more.</li>
						  <li>SNPsnap now uses the lightning fast <a href="https://www.cog-genomics.org/plink2/">PLINK 1.9</a> for clumping input SNPs.</li>
						</ul>
					We hope you enjoy the new and improved version! Please do not hesitate to <a href="feedback.html">give us feedback</a>.
					</div>
				</div> <!-- END feed -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">September 15th, 2014</div>
						<h3 class="panel-title">Let SNPsnap check the independence of your input SNPs</h3>
					</div>
					<div class="panel-body">
						Checking indepence of SNPs have never been easier: we have integrated an option for reporting the independence of your input SNPs. You can adjust the threshold of clumping (both LD and physical distance) after selecting the checkbox <span class="code_settings">Report input loci independence</span>. This will generate the file <span class="code_files">input_snps_clumped.txt</span> where researches can inspect the clumping in more details.
					</div>
				</div> <!-- END feed -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">September 11th, 2014</div>
						<h3 class="panel-title">Accepting rs-numbers</h3>
					</div>
					<div class="panel-body">
						SNPsnap now accepts rs-numbers as SNP input. In addition SNPsnap generates the file <span class="code_files">input_snps_identifer_mapping.txt</span> to ease the mapping to/from chromosomal coordinates and rs-numbers.
					</div>
				</div> <!-- END feed -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">July 14th, 2014</div>
						<h3 class="panel-title">Maching Bias</h3>
					</div>
					<div class="panel-body">
						SNPsnap now includes a matching bias calculation to determine differences between the mean genetic properties of input and matched SNPs.<br>
						This helps the user to balance the trade-off between requesting a large number of matched SNPs while avoiding systematic differences between input and matched SNPs.
						<!-- Users that request a large number of matched SNPs may find this helpful in choosing the proper matching criteria. -->
					</div>
				</div> <!-- END feed -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">July 8th, 2014</div>
						<h3 class="panel-title">Option for excluding HLA SNPs</h3>
					</div>
					<div class="panel-body">
						SNPsnap now supports a new exclusion feature <span class="code_settings">Exclude HLA SNPs</span>. <br>
						Enabling this option will exclude all input SNPs mapping to the HLA region (chr 6 25Mb-35Mb) in both input and matched SNPs. <br>
					</div>
				</div> <!-- END feed -->
				<div class="panel panel-default"> <!-- START feed -->
					<div class="panel-heading">
						<div class="text-right">June 25th, 2014</div>
						<h3 class="panel-title">SNPsnap is now airborn!</h3>
					</div>
					<div class="panel-body">
						The team behind SNPsnap is excited to announce that SNPsnap is now online. <br>
						We strongly encourage you to provide feedback for any improvements to the website or SNP matching parameters. <br>
					</div>
				</div> <!-- END feed -->
			<!-- </div> -->
		</div>
	</div>


			

 <!-- ############## FOOTER ################### -->
<div class="footer">
	<div class="col-xs-3"><p>&copy; Broad 2014</p></div>
	<div class="col-xs-8"></div>
	<div class="col-xs-1"><img class="img-responsive footer_min_limits" src="img/broad_logo/BroadSeal.png"></div>
</div>
 <!-- end FOOTER -->


</div> <!-- /container -->


</body>
</html>




<!-- 
<nav class="navbar">
		<div class="navbar-header">
				<a class="navbar-brand" href="#">
						<img style="max-width:100px; margin-top: -7px;" src="/img/broad_logo/BroadLogo.png">
				</a>

				<ul class="nav nav-pills pull-right">
					<li class="active"><a href="#">Home</a></li>
					<li><a href="#">About</a></li>
					<li><a href="#">Contact</a></li>
				</ul>
		</div>
</nav> -->

<!-- <div class="navbar-header">
		<a class="navbar-brand" rel="home" href="#" title="Buy Sell Rent Everyting">
				<img style="max-width:100px; margin-top: -7px;"
						 src="/img/broad_logo/BroadLogo.png">
		</a>
</div> -->
